File: lib/ch4/4.6.js
// Design an algorithm and write code to find the first common ancestor of
// two nodes in a binary tree Avoid storing additional nodes in a data
// structure NOTE: This is not necessarily a binary search tree

class BST {
  constructor(value) {
    this.value = value;
    this.parent = null;
    this.left = null;
    this.right = null;
  }
}

// const getHeight = node => {
//   let currNode = node;
//   let height = 1;
//   while (currNode.parent !== null) {
//     height++;
//     currNode = currNode.parent;
//   }
//   return height;
// };

// const getRoot = node => {
//   let currNode = node;
//   while (currNode.parent) {
//     currNode = currNode.parent;
//   }
//   return currNode;
// }

const lca = (root, node1, node2,) => {
  if (root === node1 || root === node2) return root;
  const left = lca(root.left, node1, node2);
  const right = lca(root.right, node1, node2);
  return (
    left === null
      ? right
      : right === null
          ? left
          : root
  );
};

File: java/ch4/4.6.java
class BTreeNode {
    BTreeNode left;
    BTreeNode right;
    int data;
    public BTreeNode(int data) {
        this.data = data;
        this.left = left;
        this.right = right;
    }
}

BTreeNode lca(BTreeNode root, BTreeNode n1, BTreeNode n2) {
    if (root == n1 || root == n2) return root;
    BTreeNode left = lca(root.left, n1, n2);
    BTreeNode right = lca(root.right, n1, n2);
    return (
        left == null
            ? right
            : right == null
                ? left
                : root
    );
}
